KAFKA-14317: ProduceRequest timeouts are logged as network exceptions

kirktrue opened the pull request:
Minor changes to `Sender` and `NetworkClient` so that we can log timeouts during `ProduceRequest` with a more precise error message, denoting a timeout vs. "generic" network error.

It appears that the meta data updater already triggers a metadata refresh, so we don't need to add that here.

See [KAFKA-14317](https://issues.apache.org/jira/browse/KAFKA-14317) for more detail.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

kirktrue committed:
KAFKA-14317: ProduceRequest timeouts are logged as network exceptions

Log timeouts during ProduceRequest with more precise error message as
well as signal the need for a metadata refresh.

dajac commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -346,8 +346,16 @@ private void cancelInFlightRequests(String nodeId, long now, Collection<ClientRe
-                if (responses != null)
-                    responses.add(request.disconnected(now, null));
+                if (responses != null) {
+                    ClientResponse disconnected;
```
nit: The variable name is not accurate anymore.

kirktrue replied:
This code has been refactored and there's no longer a new local variable set.

dajac commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -346,8 +346,16 @@ private void cancelInFlightRequests(String nodeId, long now, Collection<ClientRe
+                if (responses != null) {
+                    ClientResponse disconnected;
+
+                    if (request.timeElapsedSinceSendMs(now) > request.requestTimeoutMs)
```
I wonder if we could pass a boolean to the method instead of re-checking `timeElapsedSinceSendMs` here. Have you tried/considered this?

kirktrue replied:
Done.

dajac commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -1284,6 +1292,11 @@ public ClientResponse disconnected(long timeMs, AuthenticationException authenti
 
+        public ClientResponse timedOut(long timeMs, AuthenticationException authenticationException) {
+            return new ClientResponse(header, callback, destination, createdTimeMs, timeMs,
+                    true, true, null, authenticationException, null);
```
My understanding is that a timed out request is still considered disconnected as well. Am I correct? I think that this is important in order to not break compatibility. The NetworkClient is used in many other places.

kirktrue replied:
Correct. I have added a check in the `ClientResponse` constructor to ensure that time outs are only set when also disconnected.

dajac commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -552,9 +552,11 @@ private void handleProduceResponse(ClientResponse response, Map<TopicPartition,
-            for (ProducerBatch batch : batches.values())
-                completeBatch(batch, new ProduceResponse.PartitionResponse(Errors.NETWORK_EXCEPTION, String.format("Disconnected from node %s", response.destination())),
+            for (ProducerBatch batch : batches.values()) {
+                Errors e = response.wasTimedOut() ? Errors.REQUEST_TIMED_OUT : Errors.NETWORK_EXCEPTION;
```
I wonder if propagating the TimeoutException to the user space has any backward compatibility implications. I think that we already use TimeoutException when the batch expires so the user should already handle timeouts. That seems to be ok.

dajac commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -552,9 +552,11 @@ private void handleProduceResponse(ClientResponse response, Map<TopicPartition,
-                completeBatch(batch, new ProduceResponse.PartitionResponse(Errors.NETWORK_EXCEPTION, String.format("Disconnected from node %s", response.destination())),
+            for (ProducerBatch batch : batches.values()) {
+                Errors e = response.wasTimedOut() ? Errors.REQUEST_TIMED_OUT : Errors.NETWORK_EXCEPTION;
+                completeBatch(batch, new ProduceResponse.PartitionResponse(e, String.format("Disconnected from node %s", response.destination())),
```
nit: Should we also change the message here when the request is timed out? I also wonder if we should update the trace log at L553 as well. Perhaps, we should just add a `if (response.wasTimeOut)` before `if (response.wasDisconnected())` to fully customise it.

kirktrue replied:
I updated the patch to introduce a new `if` block for timeouts that is separate from disconnections, as you'd suggested. LMK if I interpreted your request correctly 😄

dajac commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -641,7 +643,7 @@ private void completeBatch(ProducerBatch batch, ProduceResponse.PartitionRespons
                 failBatch(batch, response, batch.attempts() < this.retries);
             }
-            if (error.exception() instanceof InvalidMetadataException) {
+            if (error.exception() instanceof InvalidMetadataException || error.exception() instanceof TimeoutException) {
```
I wonder if this is necessary. It seems that the NetworkClient calls `metadataUpdater.handleServerDisconnect` when the node is disconnected and a metadata refresh is requested there. Could this be?

kirktrue replied:
I believe it is necessary and added a comment to the code to explain why. LMK if that makes sense. Thanks!

dajac replied:
I agree that we should refresh on `TimeoutException`. I wanted to point out that we are already doing it there: https://github.com/apache/kafka/blob/d6e35f3c41880c102faf774926cfb104a012de3f/clients/src/main/java/org/apache/kafka/clients/NetworkClient.java#L786. Hence I wonder if we need to add it here as well.

kirktrue replied:
@dajac For disconnected/timed out connections, this is the call stack:

```java
Sender.poll()
    NetworkClient.poll()
        DefaultMetadataUpdater.handleServerDisconnect()
            Metadata.requestUpdate()
````

However, `SenderTest` uses `MockClient` which doesn't have that same logic flow. So if I remove that clause of the `if` statement, the test, as written, fails.

dajac replied:
@kirktrue Do we need that new test if we remove the change in the Sender? We could perhaps remove it and keep the PR focused on the initial change. This would allow us to merge that first part.

kirktrue replied:
@dajac I removed the `error.exception() instanceof TimeoutException` check in `Sender` as you suggested. The unit test for ensuring that the metadata is updated on request timeouts has been added to the pre-existing request timeout tests in `NetworkClientTest`.

dajac commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -2519,7 +2519,7 @@ public void testInflightBatchesExpireOnDeliveryTimeout() throws InterruptedExcep
         Map<TopicPartition, ProduceResponse.PartitionResponse> responseMap = new HashMap<>();
         responseMap.put(tp0, new ProduceResponse.PartitionResponse(Errors.NONE, 0L, 0L, 0L));
-        client.respond(new ProduceResponse(responseMap));
+        client.respond(new ProduceResponse(responseMap), true, true);
```
Changing this test does not seem right because it tests the delivery timeout of the batch, not the request timeout. Could we add another one?

kirktrue replied:
Ah! I reverted that change and added a dedicated `testMetadataRefreshOnRequestTimeout` test method to `SenderTest`. Let me know if that new test is sufficient.

dajac reviewed (commented):
Thanks for the patch. I think that the approach looks reasonable. We have to be careful with not breaking other usages of the NetworkClient though. Have you checked how other components use it and if everything is fine? I left a few comments.

kirktrue committed:
Updates to timeout handling

* Added check in ClientResponse constructor to make sure timedOut set only when disconnected is also set
* Passing a timedOut flag down through call chain in NetworkClient (handleTimedOutConnections and handleTimedOutRequests) vs. re-evaluating down in cancelInFlightRequests

kirktrue committed:
Added test testMetadataRefreshOnRequestTimeout

kirktrue committed:
Refactored Sender.handleProduceResponse

Refactored the logging and completeBatch calls per David's suggestion.

kirktrue committed:
MInor formatting cleanup

kirktrue committed:
Added comment for Sender.completeBatch

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -549,7 +549,13 @@ private boolean awaitNodeReady(Node node, FindCoordinatorRequest.CoordinatorType
         int correlationId = requestHeader.correlationId();
-        if (response.wasDisconnected()) {
+        if (response.wasTimedOut()) {
+            log.trace("Cancelled request with header {} due to node {} being disconnected due to timeout",
```
very nit: `due to the last request timed out`? just to distinguish from connection timeout.

kirktrue replied:
Changed to:

>Cancelled request with header {} due to the last request to node {} timed out

kirktrue committed:
Minor trace log rewording

guozhangwang commented:
I made another pass and it LGTM now. leaving to @dajac for a final look and merge.

dajac commented:
@guozhangwang Thanks. I will do this tomorrow. In the mean time, I have re-triggered the CI as they were multiple failures.

dajac commented:
@kirktrue I left a small comment here: https://github.com/apache/kafka/pull/12813#discussion_r1099905173.

kirktrue committed:
Updates to remove explicit check from Sender

The commit removes the check for TimeoutException from Sender.completeBatch. We instead rely on the fact this is already handled in NetworkClient.DefaultMetadataUpdater.handleServerDisconnect.

kirktrue committed:
Update MockClient.java

Removing the unnecessary respond method.

kirktrue committed:
Update NetworkClient.java

More code reverting.

kirktrue committed:
Update NetworkClient.java

Remove more unnecessary methods.

kirktrue committed:
Update MockClient.java

Added back the superfluous whitespace

kirktrue committed:
Update NetworkClientTest.java

Comments

philipnee commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -324,11 +324,14 @@ public void disconnect(String nodeId) {
     }
 
-    private void cancelInFlightRequests(String nodeId, long now, Collection<ClientResponse> responses) {
+    private void cancelInFlightRequests(String nodeId,
```
nit: final-lize all params

kirktrue replied:
I didn't want to introduce any inconsistency in this change as the rest of the code doesn't follow that pattern.

kirktrue replied:
I'm marking this as "resolved" for now. Please reopen in protest if you strongly disagree 😄

philipnee commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -753,7 +756,8 @@ public static AbstractResponse parseResponse(ByteBuffer responseBuffer, RequestH
     private void processDisconnection(List<ClientResponse> responses,
```
I'm thinking, instead of passing boolean flags around, would it be better to have another function to explicitly calling out `processTimeoutDisconnection` ?

kirktrue replied:
Done.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -1279,9 +1283,10 @@ public ClientResponse completed(AbstractResponse response, long timeMs) {
         }
 
-        public ClientResponse disconnected(long timeMs, AuthenticationException authenticationException) {
+        public ClientResponse disconnected(long timeMs, AuthenticationException authenticationException, boolean timedOut) {
```
and maybe we could introduce `timeouted(long timeMs, AuthenticationException authenticationException)` ? if that makes everything more explicit about the timeout situation.

kirktrue replied:
Done.

philipnee reviewed (commented):
Thanks! Overall it makes a lot of sense to me so I left some coding structure comnents.

kirktrue committed:
Introduced an overloaded processDisconnection and processTimeoutDisconnection

kirktrue committed:
InFlightRequest method cleanup

kirktrue commented:
@dajac @philipnee I believe I've addressed the questions/comments/concerns that were raised. There are two test failures that seem unrelated. Is there anything blocking this from being merged? Thanks!

philipnee commented:
Thanks Kirk! ✅

dajac commented:
Thanks. I will take a look and merge it in the next few days.

dajac reviewed (approved):
LGTM, thanks.
